### ISSUE TITLE
plot.overlay_href does not return the layer

### ISSUE THREAD

**Issue by mrecachinas** [LGSInnovations/sigplot#32]
**Describe the bug**
[`plot.overlay_href`](https://github.com/LGSInnovations/sigplot/blob/master/js/sigplot.js#L2393-L2432) does not return the layer that was created (despite the docs saying it does).

**To Reproduce**

```javascript
var layer = plot.overlay_href('dat/penny.prm', null, {});
console.log(layer); // => undefined
```

**Expected behavior**
`plot.overlay_href` should return the layer index for the layer that was created.

**jsFiddle**

N/A

**Screenshots**
N/A

**Desktop (please complete the following information):**
N/A

**Smartphone (please complete the following information):**
N/A

**Additional context**
```javascript
        overlay_href: function(href, onload, layerOptions) {
            m.log.debug("Overlay href: " + href);
            try {
                this.show_spinner();

                var handleHeader = (function(plot, onload) {
                    return function(hcb) {
                        try {
                            if (!hcb) {
                                alert("Failed to load data");
                            } else {
                                var i;
                                if (href.endsWith(".mat")) {
                                    i = plot.overlay_matfile(hcb, layerOptions);
                                } else {
                                    i = plot.overlay_bluefile(hcb, layerOptions);
                                }
                                if (onload) {
                                    onload(hcb, i);
                                }
                            }
                        } finally {
                            plot.hide_spinner();
                        }
                    };
                }(this, onload));

                var reader;
                if (href.endsWith(".mat")) {
                    reader = new matfile.MatFileReader();
                } else {
                    reader = new bluefile.BlueFileReader();
                }
                reader.read_http(href, handleHeader);
            } catch (error) {
                console.error(error);
                alert("Failed to load data");
                this.hide_spinner();
            }
        },
```

**Comment by mrecachinas** [comment]
After discussing this with @maihde, it appears this is intentional because of the asynchronous nature of file-loading.

For the time being, this issue can be solved by modifying the docs.

The longer term solution might include returning the layer index of where the layer _will_ be once the file is loaded.

**Comment by sterre** [comment]
Would it be helpful for `overlay_href` to take an optional callback argument, triggered when the layer is added sometime in the future? Or maybe SigPlot could trigger some other event when a layer resolves?

**Comment by mrecachinas** [comment]
`overlay_href` does take in an optional callback triggered when the layer is added.

**Comment by sterre** [comment]
> `overlay_href` does take in an optional callback triggered when the layer is added and you have access to the layer index within the callback.

Then clearly it was a really good idea and all of us who thought of it independently are Quality People. I do like the idea of being able to record immediately where the layer *will* be, for future access; but given that you won't know the layer is really there until after the callback fires, I don't see any reason to hurry.

**Comment by mrecachinas** [comment]
@sterre right -- and that's fair. This came up with bookkeeping within react-sigplot for the HrefLayer. I believe I've found a way around it (using the callback), but it is an open question.

**Comment by desean1625** [comment]
Would it make sense to change all the overlay functions to return a promise? This would allow you to write await functions for all the overlay methods.

**Comment by mrecachinas** [comment]
@desean1625 This would definitely be an API breaking change, but I agree it makes a lot of sense. For this big of a change, definitely going to want @maihde to weigh in.

**Comment by desean1625** [comment]
Or a non breaking change only return the promise on overlay_href.On May 16, 2019 5:18 PM, Michael Recachinas <notifications@github.com> wrote:@desean1625 This would definitely be an API breaking change, but I agree it makes a lot of sense. For this big of a change, definitely going to want @maihde to weigh in.

**Comment by mrecachinas** [comment]
Given 45778ab20c9495d5334e1c97fd7e484e56b2267f, where multiple hrefs can now be plotted with `overlay_href`, should we return a list of promises (i.e., one promise per layer)? A promise where a parameter is a list of layers?

**Comment by maihde** [comment]
@mrecachinas @desean1625 @sterre great news...I actually got this implemented last night.  If users were treating the return value from `overlay_bluefile` and friends as an opaque value then there are no API breaks.  If someone was ignoring the return value and hardcoding layer numbers into their `push` or `reload` calls then they will need to change their code.  I'll submit a PR soon...but you can check out:

https://github.com/spectriclabs/sigplot/commit/3167fef123e8f2e1769ed657cb0f50b679448300
https://github.com/spectriclabs/sigplot/commit/52e8178d6d7345be6378d12fa87159cee9b2cb59
https://github.com/spectriclabs/sigplot/commit/a7d94bd77e4da8ac30b91460b10fa785020e3596

**Comment by maihde** [comment]
Here is a PR https://github.com/LGSInnovations/sigplot/pull/89

**Comment by mrecachinas** [comment]
Fixed in #89.
